app/app.go

```diff
@@ -1121,9 +1129,43 @@ func (app *App) ProcessBlockConcurrent(
 	return txResults, ok
 }
 
+func (app *App) ProcessTxs(
+	ctx sdk.Context,
+	txs [][]byte,
+	dependencyDag *acltypes.Dag,
+	processBlockConcurrentFunction ProcessBlockConcurrentFunction,
+) []*abci.ExecTxResult {
+	// Only run concurrently if no error
+	// Branch off the current context and pass a cached context to the concurrent delivered TXs that are shared.
+	// runTx will write to this ephermeral CacheMultiStore, after the process block is done, Write() is called on this
+	// CacheMultiStore where it writes the data to the parent store (DeliverState) in sorted Key order to maintain
+	// deterministic ordering between validators in the case of concurrent deliverTXs
+	processBlockCtx, processBlockCache := app.CacheContext(ctx)
+	concurrentResults, ok := processBlockConcurrentFunction(
+		processBlockCtx,
+		txs,
+		dependencyDag.CompletionSignalingMap,
+		dependencyDag.BlockingSignalsMap,
+		dependencyDag.TxMsgAccessOpMapping,
+	)
+	if ok {
+		// Write the results back to the concurrent contexts - if concurrent execution fails,
+		// this should not be called and the state is rolled back and retried with synchronous execution
+		processBlockCache.Write()
+		return concurrentResults
+	}
+
+	ctx.Logger().Error("Concurrent Execution failed, retrying with Synchronous")
+	// Clear the memcache context from the previous state as it failed, we no longer need to commit the data
+	ctx = ctx.WithContextMemCache(sdk.NewContextMemCache())
```

Comment:
just to confirm, is this the step that we need in order to reset the in memory state prior to rerunning txs synchronously?

Reply:
Yes, this resets the state for the rerun of TXs synchronously